bip-0181.md

```diff
@@ -0,0 +1,647 @@
+08      09      10      11
+|---\   |---\   |---\   |---\
+00  01  02  03  04  05  06  07
+```
+
+The new accumulator will look like so:
+
+```
+
+|-------------------------------\
+28
+|---------------\               |---------------\
+24              25
+|-------\       |-------\       |-------\       |-------\
+16      17      18      19
+|---\   |---\   |---\   |---\   |---\   |---\   |---\   |---\
+00  01  02  03  04  05  06  07  08
+```
+
+The new accumulator with all the positions:
+
+```
+30
+|-------------------------------\
+28                              29
+|---------------\               |---------------\
+24              25              26              27
+|-------\       |-------\       |-------\       |-------\
+16      17      18      19      20      21      22      23
+|---\   |---\   |---\   |---\   |---\   |---\   |---\   |---\
+00  01  02  03  04  05  06  07  08  09  10  11  12  13  14  15
+```
+
+# Definitions
+
+- `hash` refers to a vector of 32 byte arrays.
+- `[]hash` refers to a vector of `hash`.
+- `acc` refers to the Utreexo accumulator state. An `acc` is comprised of:
+  - `roots` refers to the roots of the Merkle Trees. Represented as `[]hash`.
+  - `numleaves` refers to the number of total leaves added to the accumulator. Represented as uint64.
+- `root` refers to the top `hash` in a tree in the `acc`.
+- `proof` is an inclusion proof for elements in the accumulator. It's comprised of two fields:
+  - `targets` are the positions of the elements being proven. Represented as a vector of uint64.
+  - `proof` are the hashes needed to hash the roots. Represented as a `[]hash`. `proof` MUST be in ascending order by the node positions.
+    The proof is considered invalid otherwise.
+
+# Specification
+
+The hash function SHA512/256[^2] is used for the hash operations in the accumulator.
+
+An Utreexo accumulator implementation MUST support these 3 operations: Addition, Verification, and Deletion.
+
+## Utility Functions
+
+The following utility functions are required for performing accumulator operations:
+
+**parent_hash(left, right):** Returns the hash of the concatenation of two child hashes (`left` and `right`).
+If either child is `nil`, the result is simply the non-`nil` child (treated as if the tree has a single child at that position).
+if both children are `nil`, the result is `nil`.
+
+Implementation:
+
+```python
+def parent_hash(left: bytes, right: bytes) -> bytes:
+    if right is None and left is None: return None
+    if left is None: return right
+    if right is None: return left
+
+    return sha512_256(left + right)
+```
+
+**treerows(numleaves):** Returns the minimum number of bits required to represent `numleaves - 1`. This corresponds to the height of the largest tree in the forest. Returns `0` if `numleaves` is `0`.
+
+The reason for taking the minimum number of bits required for `numleaves-1` and not `numleaves` is because when `numleaves` is a power of two, we'd get an off-by-one error.
```

Comment:
It would be nice to have consistent spacing around the minus sign, there's both `numleaves - 1` and `numleaves-1`. Same goes for the equals sign below.